### DISCUSSION TITLE
Multiple POST endpoints generated for one model bound action method.

### DISCUSSION THREAD

**Discussion by sharathgt** [OData/AspNetCoreOData#693] [Q&A]
I have a POST action which is bound to the model. This action sure falls into Odata endpoints list with the name .../Agencies/Default.AgencyDashboard. But It also takes other POST calls that is; POST: .../Agencies (No method)
Whenever I tried to POST:.../Agencies with a agency payload, it hits Default.AgencyDashboard.

So now, In reality I have 2 POST Actions, but 6 ODataEndPoints...

![image](https://user-images.githubusercontent.com/12605026/189275730-6b609ff9-d3a5-4666-9a46-f0ddfba369b9.png)

@julealgon,
For now to make it work, I bound the SaveAgency method also. It works now but, how to get rid of these extra endpoints.

```
[HttpPost]
public async Task<ActionResult<Agency>> AddAgency(ODataActionParameters parameters)
{
}

[HttpPost]
public async Task<ActionResult> AgencyDashboardTopbar(ODataActionParameters parameters)
{
}

[HttpPost]
public async Task<ActionResult> AgencyDashboard(ODataActionParameters parameters)
{
}
```
And the binding lines are below;

```
    var fun = builder.EntityType<Agency>().Collection.Action("AgencyDashboard");
    fun.Parameter<long>("key"); 
    fun.Parameter<int>("month").Optional();
    fun.Parameter<int>("year").Optional();
    fun.Returns<Task<ActionResult>>();

    var funG = builder.EntityType<Agency>().Collection.Action("AddAgency");
    funG.Parameter<Agency>("agency");
    funG.Returns<Task<ActionResult>>();

    var fun4 = builder.EntityType<Agency>().Collection.Action("AgencyDashboardTopbar");
    fun4.Parameter<long>("key");
    fun4.Returns<Task<ActionResult>>();

```

It is creating 3 different POST for one bound method. 

I don't know what's going wrong. Is it a bug or am I missing something.

**Comment by julealgon** [comment]
Yeah.... honestly, this is kind of a mess.

What is happening, is that 2 of the routes are being generated by conventions (the "correct" ones, "Default.ActionName" (includes the namespace) and "ActionName" (excludes the namespace)), while the other one at the root is picking up your `[HttpPost]` attribute and using the empty route there as yet another (totally invalid...) route.

As far as I'm aware, the only way to take full control over this mess is to rely _solely_ on attribute routing (by specifying the whole route template explicitly, like `[Post("AgencyDashboardTopbar")]`, and then _disable the standard conventions_, by removing them from the `options.Conventions` collection during the startup.

Removing the conventions is a big mess in and of itself. Unfortunately, they can't just be "disabled", and you can't just _clear_ the whole convention collection otherwise you also lose the `$metadata` endpoint.

**Comment by sharath-ots** [comment]
@julealgon ,

With lots of circus, moved all the non-Odata endpoint mappings to Odata endpoints; by binding, by renaming, by moving to different controllers, by changing the way to send parameters(ODataActionParameters), etc.

Struggled but could move them all. But still have few issues, which eventually I hope I get hold of them; just for information
1. The naming and thus duplicates still there... there are two endpoints with Default. and without that.
2. The response for some requests comes with @odata.context and a value like below
  
```
{
    "@odata.context": "https://.../api/v2/$metadata#Agencies",
    "value":[{},{}]
}
```
and sometimes not.
3. When it comes with @odata.context, the properties are PascalCased and when it doesn't, they will be camelCased. (Not having Uniformity in the way response is built, which is not good for frontend)

Thanks again for your valuable time.

**Reply by julealgon** [reply]
> and sometimes not.

These are probably not being recognized as proper OData endpoints.

Can you paste your results from the `$odata$ route debug endpoint?

> 3. When it comes with @odata.context, the properties are PascalCased and when it doesn't, they will be camelCased. (Not having Uniformity in the way response is built, which is not good for frontend)

This is because those non-OData payloads are being serialized with the standard MVC serializer (System.Text.Json on newer frameworks), while the OData payloads have a custom serializer.

Each of those is independent of the other

**Comment by sharath-ots** [comment]
It's difficult to bring two POSTs or two GETs in same controller, even with different name or parameters.

I feel, the whole thing very fragile... I understand different conventions and their interferences, but we also have Order... there must be a simple and "predictable" way of doing this. This is very uncertain, unpredictable for a newcomer...

The list is big... it goes like this... and I do not have any non-OData mappings now

![image](https://user-images.githubusercontent.com/113424437/190219419-befa4578-9207-41a5-98a6-b37af8fa052b.png)

![image](https://user-images.githubusercontent.com/113424437/190219535-72594e1c-0dd8-435e-8ae2-97bfe428069f.png)

**Reply by julealgon** [reply]
I totally agree with you that this is very developer-unfriendly right now, especially if you don't know the rules precisely (which the vast majority of people will not).

At this point, I think the standard convention should either be dropped, or have to be turned on explicitly, and then ask folks to focus on attribute routing by default.

**Comment by vigouredelaruse** [comment]
this is mostly a response to notice that this is happening to others, and to say
* i feel your pain and respond to it in my own circumstances with an ACHTUNG.md in the OData controller folder, something like https://dev.azure.com/electricrucible/_git/the%20horseless%20newspaper?path=/src/core/TheHorselessNewspaper/HorselessNewspaper.RazorClassLibrary.CMS.Default/HorselessControllers/OData/ACHTUNG.md
* i suspect perceived fragility is related to a lack of visibility into the behaviour of route resolution in asp.net core - the route endpoints really correspond to something like a priority based firewall ruleset and it's necessary to instrument path resolution failures in order to eliminate them, sometimes with this https://github.com/ardalis/AspNetCoreRouteDebugger
